pkg/gds/admin.go

```diff
@@ -149,22 +160,224 @@ func (s *Admin) setupRoutes() (err error) {
+// provided in the Authorization header as a Bearer token, even if it is expired since
+// the access token contains the claims that need to be reissued. The refresh token is
+// posted in the request body as the credential. This method also resets the CSRF double
+// cookies to ensure that the max-age matches the duration of the refresh tokens.
 func (s *Admin) Reauthenticate(c *gin.Context) {
-	if err := admin.SetDoubleCookieTokens(c, time.Now().Add(time.Minute*10).Unix()); err != nil {
+	var (
+		err           error
+		tks           string
+		in            *admin.AuthRequest
+		out           *admin.AuthReply
+		accessClaims  *tokens.Claims
+		refreshClaims *tokens.Claims
+		accessToken   *jwt.Token
+		refreshToken  *jwt.Token
+	)
+
+	// Get the Bearer token from the Authorization header (contains access token)
+	if tks, err = admin.GetAccessToken(c); err != nil {
+		log.Warn().Err(err).Msg("reauthenticate called without access token")
+		c.JSON(http.StatusUnauthorized, admin.ErrorResponse("request is not authorized"))
+		return
+	}
+
+	// Parse the access token from the Authorization header without validating the
+	// claims, e.g. it doesn't matter if the access token is expired, but it should be
+	// signed correctly by the token server.
+	if accessClaims, err = s.tokens.Parse(tks); err != nil {
+		log.Warn().Err(err).Msg("reauthenticate called with invalid access token")
+		c.JSON(http.StatusUnauthorized, admin.ErrorResponse("request is not authorized"))
+		return
+	}
+
+	// Parse incoming JSON data from the client request (contains refresh token)
+	in = new(admin.AuthRequest)
+	if err = c.ShouldBind(&in); err != nil {
+		log.Warn().Err(err).Msg("could not bind request")
+		c.JSON(http.StatusBadRequest, admin.ErrorResponse(err))
+		return
+	}
+
+	// Check that a credential was posted
+	if in.Credential == "" {
+		c.JSON(http.StatusUnauthorized, admin.ErrorResponse("invalid credentials"))
+		return
+	}
+
+	// Validate the refresh token
+	if refreshClaims, err = s.tokens.Verify(in.Credential); err != nil {
+		log.Warn().Err(err).Msg("could not verify refresh token")
+		c.JSON(http.StatusUnauthorized, admin.ErrorResponse("invalid credentials"))
+		return
+	}
+
+	// Ensure the refresh token and admin token match
+	// TODO: verify the in.Credential is a refresh token using the subject or audience
+	if accessClaims.Id != refreshClaims.Id {
+		log.Warn().Msg("mismatched access and refresh token pair")
+		c.JSON(http.StatusUnauthorized, admin.ErrorResponse("invalid credentials"))
+		return
+	}
+
+	// At this point we've validated the reauthentication and are ready to reissue tokens
+	// Create the access and refresh tokens from the claims
+	if accessToken, err = s.tokens.CreateAccessToken(accessClaims); err != nil {
+		log.Error().Err(err).Msg("could not create access token")
+		c.JSON(http.StatusInternalServerError, admin.ErrorResponse("could not authenticate with credentials"))
```

Comment:
What are some of the scenarios that would cause reauthentication to fail on line 347 if the access and refresh tokens comparison succeeds on 339?

Do we want our log messages on lines 348, 354, 362, 367 to reflect that this is a reauthentication fail vs a first-time authentication fail?

Reply:
If we get to line 347 then something is wrong on the server; e.g. the token RSA keys are bad, something is wrong with the configuration, or a developer error occurred. 348, 354, 362, and 667 return an 500 error (something went really wrong) instead of a 401 error (you're not authorized) and the log level increases from `Warn()` to `Error()`, and hopefully we'll receive the Slack alert and fix it! 

So to answer your question - it would be a pretty narrow edge case, for example the user authenticates then we restart the server with a new configuration and then the user attempts to reauthenticate. However, your point about messaging which method we're in is a good one as that may help debugging - so I'll changing the wording!